File: src/main/java/com/jacuum/engine/ActiveSession.java
package com.jacuum.engine;

import com.jacuum.map.GameMap;
import java.util.HashSet;
import java.util.Set;
import java.util.concurrent.Future;

final class ActiveSession {
    final String id;
    final GameMap map;
    final String algoName;
    final String username;
    final String avatar;
    final int iterationsAvailable;

    volatile int robotX;
    volatile int robotY;
    volatile int score;
    volatile int iterationsUsed;
    volatile RunStatus status;
    volatile FinishReason finishReason;
    final Set<String> cleaned = new HashSet<>();
    volatile Future<?> future;

    ActiveSession(String id, GameMap map, String algoName,
                  String username, String avatar, int iterationsAvailable) {
        this.id                   = id;
        this.map                  = map;
        this.algoName             = algoName;
        this.username             = username;
        this.avatar               = avatar;
        this.iterationsAvailable  = iterationsAvailable;
        this.robotX               = map.startX();
        this.robotY               = map.startY();
        this.status               = RunStatus.SETUP;
    }

    SessionView toView() {
        return new SessionView(id, status, robotX, robotY, score,
            cleaned.size(), iterationsUsed, iterationsAvailable,
            map.totalFloorTiles(), finishReason);
    }
}

File: src/main/kotlin/com/jacuum/engine/ActiveSession.kt
package com.jacuum.engine

import com.jacuum.map.GameMap
import kotlinx.coroutines.Job
import kotlinx.coroutines.sync.Mutex
import java.util.concurrent.ConcurrentHashMap

internal class ActiveSession(
    val id: String,
    val map: GameMap,
    val algoName: String,
    val username: String,
    val avatar: String,
    val iterationsAvailable: Int
) {
    @Volatile var robotX: Int = map.startX()
    @Volatile var robotY: Int = map.startY()
    @Volatile var score: Int = 0
    @Volatile var iterationsUsed: Int = 0
    @Volatile var status: RunStatus = RunStatus.SETUP
    @Volatile var finishReason: FinishReason? = null
    val cleaned: MutableSet<String> = ConcurrentHashMap.newKeySet()
    var job: Job? = null
    // Unlocked = running; locked = paused. Game loop calls withLock{} to suspend while paused.
    val pauseGate: Mutex = Mutex()

    fun toView() = SessionView(
        id, status, robotX, robotY, score,
        cleaned.size, iterationsUsed, iterationsAvailable,
        map.totalFloorTiles(), finishReason
    )
}
